Enable replication of `krblastsuccessfulauth` timestamps

jsf9k opened the pull request:
## 🗣 Description ##

This pull request adds a `bash` script to modify the replication agreements between FreeIPA servers to enable replication of `krblastsuccessfulauth` timestamps.  [By default these timestamps *are not* replicated](https://pagure.io/freeipa/issue/9821) between servers to avoid a replication storm, but given our limited number of users we should be OK.

While I was here I went ahead and broke up `01_setup_disabling_of_inactive_freeipa_users.sh` into several functions, and modified the script so that it is possible to run a single function at a time.  The default behavior remains the same, but it is useful to have the ability to run individual functions sometimes.

## 💭 Motivation and context ##

Resolves #82.

## 🧪 Testing ##

All automated tests pass.  As part of the testing for cisagov/freeipa-server-packer#146 I also deployed these changes to our `dev-a` and `staging-a` COOL environments and verified that there they function as expected.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All new and existing tests pass.
- [x] Bump major, minor, patch, pre-release, and/or build versions [as appropriate](https://semver.org/#semantic-versioning-specification-semver) via the `bump_version` script *if* this repository is versioned *and* the changes in this PR [warrant a version bump](https://semver.org/#what-should-i-do-if-i-update-my-own-dependencies-without-changing-the-public-api).
- [x] Create a pre-release (necessary if and only if the pre-release version was bumped).

## ✅ Pre-merge checklist ##

- [x] Finalize version.

## ✅ Post-merge checklist ##

- [x] Create a release (necessary if and only if the version was bumped).

jsf9k committed:
Use systemctl enable --now to force timer start

jsf9k committed:
Break bash code out into individual functions

The code was getting a bit too long and difficult to read, and this
helps.

jsf9k committed:
Add a new script to enable replication of krblastsuccessfulauth timestamps

By default these timestamps *are not* replicated between servers to
avoid a replication storm, but given our limited number of users we
should be OK.  See here for more details:
https://pagure.io/freeipa/issue/9821

jsf9k committed:
Bump version from 2.0.0 to 2.1.0-rc.1

jsf9k committed:
Update script to allow calling of a particular function

This is occasionally useful if a part of the setup process fails.

jsf9k committed:
Bump version from 2.1.0-rc.1 to 2.1.0-rc.2

jsf9k committed:
Add quotes around string value that contains a space

jsf9k committed:
Also remove extra space character in parameter expansions

jsf9k committed:
Bump version from 2.1.0-rc.2 to 2.1.0-rc.3

dav3r reviewed (approved):
LGTM 💪 💼

mcdonnnj commented on files/02_amend_replication_agreements.sh:
```diff
@@ -0,0 +1,53 @@
+      # modify it.
+      cmd_output=$(ipa topologysegment-find "$suffix" --all \
+        --name="$segment" --raw)
+      old_repl_attr=$(sed --quiet \
+        "s/^[[:blank:]]*nsDS5ReplicatedAttributeList:[[:blank:]]*\(.*\)$/\1/p" \
+        <<< "$cmd_output")
+      new_repl_attr=${old_repl_attr// krblastsuccessfulauth/}
+      old_repl_attr_total=$(sed --quiet \
+        "s/^[[:blank:]]*nsDS5ReplicatedAttributeListTotal:[[:blank:]]*\(.*\)$/\1/p" \
+        <<< "$cmd_output")
+      new_repl_attr_total=${old_repl_attr_total// krblastsuccessfulauth/}
```
It might be nice to add a brief comment to make it clear that this is where we remove `krblastsuccessfulauth` from the list of replication attributes.

jsf9k replied:
Please see commit 97f223a.

mcdonnnj commented on files/01_setup_disabling_of_inactive_freeipa_users.sh:
Please mention the work on this script in the pull request description for documentation since I do not see the connection with `krblastsuccessfulauth` timestamp replication.

jsf9k replied:
Please see the (revised) PR description.

mcdonnnj commented on files/01_setup_disabling_of_inactive_freeipa_users.sh:
```diff
@@ -6,11 +6,16 @@ set -o pipefail
-  # Disable the "KDC:Disable Last Success" password plugin feature.
-  # We don't want this feature enabled because we want to be able to
-  # disable inactive users, which requires us to be able to determine
-  # the last time a user authenticated.
-  #
+  disable_password_plugin_feature
+  create_role_and_privilege
+  enable_systemd_timer
+}
```
A small nitpick but I'd prefer this function declaration to be the last one since it uses the three other functions.

jsf9k replied:
Please see commit b0286c8.

mcdonnnj reviewed (approved):
This all LGTM ✔ I have just a few minor feedback items around formatting/documentation.

jsf9k committed:
Add comments explaining where we are removing excluded replication attributes

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Move setup function so that it comes after the other functions

This makes sense because it uses all the other functions.

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Improve comment

jsf9k committed:
Correct command to re-enable errexit

jsf9k committed:
Update script to allow calling of a particular function

This is occasionally useful if a part of the script fails on the first
try.

jsf9k committed:
Add code to reinitialize this replica

This is necessary for the changes made to the topology segment
replication agreements to take effect immediately.

jsf9k committed:
Rename script so that the name better describes what it does

jsf9k committed:
Bump version from 2.1.0-rc.3 to 2.1.0-rc.4

jsf9k committed:
Fix typo in command

jsf9k committed:
Use read to convert the topology suffix string to a bash array

This is a more robust way of extracting the first topology suffix.

jsf9k committed:
Tweak comment

jsf9k committed:
Fix sed command so that it only prints the first match

The way I was trying to do it previously did not work, but this way
should.  The T means "jump if no substitution took place" which means
that the p and q commands will ONLY run if a substitution took place;
otherwise, we proceed to the next line.

jsf9k committed:
Bump version from 2.1.0-rc.4 to 2.1.0-rc.5

jsf9k committed:
Fix typo in command

The -d flag was added in error.

jsf9k committed:
Handle case where the topology segment has a name that is backwards

In other words, where the replica where the script is being run is the
left node.

jsf9k committed:
Bump version from 2.1.0-rc.5 to 2.1.0-rc.6

jsf9k committed:
Change TOPOLOGY_SUFFIXES to a bash array

This greatly simplifies the extraction of the first topology suffix in
the reinitialize_replica function.

jsf9k committed:
Temporarily unset errexit

This accounts for the fact that the first ipa topologysegment-find|sed
command may fail if the topology segment is the other way around in
the topology segment we're looking for, i.e., if the server where
we're running this script is the left node in the topology segment.

jsf9k committed:
Improve some comments

jsf9k committed:
Bump version from 2.1.0-rc.6 to 2.1.0-rc.7

dav3r commented on files/02_amend_replication_agreements_and_reinitialize.sh:
```diff
@@ -0,0 +1,119 @@
+  # This command may fail if the topology segment we're looking for is
+  # the other way around, i.e., if the server where we're running this
+  # script is the left node in the topology segment.  In that case
+  # segment_name will be empty and we will run a different command
```
Sorry for the nitpick, but you have periods everywhere else.

```suggestion
  # segment_name will be empty and we will run a different command.
```

dav3r reviewed (approved):
Strong work! 💪

mcdonnnj commented on files/02_amend_replication_agreements_and_reinitialize.sh:
```diff
@@ -0,0 +1,119 @@
+      cmd_output=$(ipa topologysegment-find "$suffix" --all \
+        --name="$segment" --raw)
+      # Extract the part where the fractional replication attributes
+      # are specified
```
Consistency with periods.

```suggestion
      # are specified.
```

mcdonnnj commented on files/02_amend_replication_agreements_and_reinitialize.sh:
```diff
@@ -0,0 +1,119 @@
+        "s/^[[:blank:]]*nsDS5ReplicatedAttributeList:[[:blank:]]*\(.*\)$/\1/p" \
+        <<< "$cmd_output")
+      # Remove krblastsuccessfulauth from the list of excluded
+      # fractional replication attributes
```
Consistency with periods.

```suggestion
      # fractional replication attributes.
```

mcdonnnj commented on files/02_amend_replication_agreements_and_reinitialize.sh:
```diff
@@ -0,0 +1,119 @@
+      # fractional replication attributes
+      new_repl_attr=${old_repl_attr// krblastsuccessfulauth/}
+      # Extract the part where the total replication attributes are
+      # specified
```
Consistency with periods.

```suggestion
      # specified.
```

mcdonnnj commented on files/02_amend_replication_agreements_and_reinitialize.sh:
```diff
@@ -0,0 +1,119 @@
+        "s/^[[:blank:]]*nsDS5ReplicatedAttributeListTotal:[[:blank:]]*\(.*\)$/\1/p" \
+        <<< "$cmd_output")
+      # Remove krblastsuccessfulauth from the list of excluded total
+      # replication attributes
```
Consistency with periods.

```suggestion
      # replication attributes.
```

mcdonnnj commented on files/02_amend_replication_agreements_and_reinitialize.sh:
```diff
@@ -0,0 +1,119 @@
+  segment_name=$(ipa topologysegment-find "$first_suffix" --pkey-only \
+    --rightnode="$my_hostname" \
+    | sed --quiet "s/^[[:blank:]]*Segment name:[[:blank:]]*\(.*\)$/\1/; T; p; q")
+  set +o errexit
```
Did you mean to re-enable it?

```suggestion
  set -o errexit
```

mcdonnnj reviewed (approved):
💪💪💪 Thanks for your continued work on this functionality. I have some minor feedback items for your consideration.

jsf9k committed:
Add periods where needed for consistency

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>
Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
Fix typo where errexit was accidentally left disabled

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Finalize version from 2.1.0-rc.7 to 2.1.0
